README.md
# AceSorting

Various sorting functions targeted for the Arduino environment, implemented
using C++11 template functions. The library is intended to be
compatible with most Arduino platforms, but some tradeoffs lean in favor of
lower-end processors (e.g. AVR, lower-end STM32) as opposed to higher-end
processors with larger amounts of RAM (e.g. ESP32, higher-end STM32).

Supports the following algorithms:

* Bubble Sort
    * `bubbleSort()` (not recommended)
* Insertion Sort
    * `insertionSort()` (recommended if N < ~100 or a stable sort is needed)
* Selection Sort
    * `selectionSort()` (not recommended)
* Shell Sort
    * `shellSortClassic()`: gap factor 2
    * `shellSortKnuth()`: gap factor 3 (recommended)
    * `shellSortTokuda()`: gap factor 2.25
* Comb Sort
    * `combSort13()`: gap factor 1.3 (13/10)
    * `combSort13m()`: gap factor 1.3, modified for gaps 9 and 10 (recommended
      for 32-bit processors)
    * `combSort133()`: gap factor 1.33 (4/3) (recommended for 8-bit processors)
    * `combSort133m()`: gap factor 1.33, modified for gaps 9 and 10
* Quick Sort
    * `quickSortMiddle()`: pivot on middle element (recommended)
    * `quickSortMedian()`: pivot on median of low, mid, high
    * `quickSortMedianSwapped()`: pivot on median and swap low, mid, high

**tl;dr**

* In most cases, use `shellSortKnuth()`.
    * It costs only 142 bytes on an AVR and 80-112 bytes on 32-bit processors.
    * It is faster than any `O(N^2)` algorithm while consuming only 34-82 extra
      bytes of flash over `insertionSort()`.
* If `N >= ~1000`, *and* you have sufficient static memory for recursive
  functions, *and* `shellSortKnuth()` is not fast enough, use:
    * `quickSortMiddle()` on 8-bit AVR processors, and
    * `quickSortMedianSwapped()` on 32-bit processors.
* Use `combSort133()` or `shellSortClassic()` to get the smallest sorting
  function faster than `O(N^2)`.
* Use `insertionSort()` if you need a stable sort.
* Don't use the C library `qsort()`.
    * It is 2-3X slower than the `quickSortXxx()` functions in this library, and
      consumes 4-5X more in flash bytes.
* Never use Bubble Sort.
    * Insertion Sort is 5-6X faster with only a handful (0-32) bytes of extra
      flash.

**Version**: 1.0.0 (2021-12-04)

**Changelog**: [CHANGELOG.md](CHANGELOG.md)

## Table of Contents

* [Hello Sorting](#HelloSorting)
* [Installation](#Installation)
    * [Source Code](#SourceCode)
    * [Dependencies](#Dependencies)
* [Documentation](#Documentation)
    * [Examples](#Examples)
* [Usage](#Usage)
    * [Include Header and Namespace](#HeaderAndNamespace)
    * [Bubble Sort](#BubbleSort)
    * [Insertion Sort](#InsertionSort)
    * [Selection Sort](#SelectionSort)
    * [Shell Sort](#ShellSort)
    * [Comb Sort](#CombSort)
    * [Quick Sort](#QuickSort)
    * [C Library Qsort](#CLibraryQsort)
* [Advanced Usage](#AdvancedUsage)
    * [Function Pointer](#FunctionPointer)
    * [Lambda Expression](#LambdaExpression)
    * [Compiler Optimizations](#CompilerOptimizations)
* [Resource Consumption](#ResourceConsumption)
    * [Flash And Static Memory](#FlashAndStaticMemory)
    * [CPU Cycles](#CpuCycles)
* [System Requirements](#SystemRequirements)
    * [Hardware](#Hardware)
    * [Tool Chain](#ToolChain)
    * [Operating System](#OperatingSystem)
* [Bugs and Limitations](#BugsAndLimitations)
* [Alternative Libraries](#AlternativeLibraries)
* [License](#License)
* [Feedback and Support](#FeedbackAndSupport)
* [Authors](#Authors)

<a name="HelloSorting"></a>
## Hello Sorting

This is a simplified version of the
[examples/HelloSorting](examples/HelloSorting) demo:

```C++
#include <Arduino.h>
#include <AceSorting.h>

using ace_sorting::shellSortKnuth;

const uint16_t ARRAY_SIZE = 20;
int array[ARRAY_SIZE];

void printArray(int* array, uint16_t arraySize) {
  for (uint16_t i = 0; i < arraySize; i++) {
    Serial.print(array[i]);
    Serial.print(' ');
  }
  Serial.println();
}

void fillArray(int* array, uint16_t arraySize) {
  for (uint16_t i = 0; i < arraySize; i++) {
    array[i] = random(256);
  }
}

void setup() {
  delay(1000);
  Serial.begin(115200);
  while (!Serial); // Leonardo/Micro

  // Attempt to get a random seed using the floating analog pin.
  randomSeed(analogRead(A0));

  fillArray();
  Serial.println("Unsorted:");
  printArray(array, arraySize);

  // The compiler automatically generates the correct version of
  // shellSortKnuth() based on the type of `array`.
  shellSortKnuth(array, arraySize);
  Serial.println("Sorted:");
  printArray(array, arraySize);
}

void loop() {}
```

<a name="Installation"></a>
## Installation

The latest stable release is available in the Arduino IDE Library Manager.
Search for "AceSorting". Click install.

The development version can be installed by cloning the
[GitHub repository](https://github.com/bxparks/AceSorting), checking out the
`develop` branch, then manually copying over to or creating a symlink from the
`./libraries` directory used by the Arduino IDE. (The result is a directory
or link named `./libraries/AceSorting`.)

The `master` branch contains the stable releases.

<a name="SourceCode"></a>
### Source Code

The source files are organized as follows:
* `src/AceSorting.h` - main header file
* `src/ace_sorting/` - implementation files
* `tests/` - unit tests which require [AUnit](https://github.com/bxparks/AUnit)
* `examples/` - example sketches
* `docs/` - contains the doxygen docs and additional manual docs

<a name="Dependencies"></a>
### Dependencies

This library itself does *not* depend on any other library.

The unit tests depend on:

* AUnit (https://github.com/bxparks/AUnit)

Some of the examples may depend on:

* AceCommon (https://github.com/bxparks/AceCommon)

<a name="Documentation"></a>
## Documentation

* this `README.md` file
* [Doxygen docs](https://bxparks.github.io/AceSorting/html) on Github pages.

<a name="Examples"></a>
### Examples

* [examples/HelloSorting](examples/HelloSorting)
    * A simple demo of one of the sorting functions.
* [examples/CompoundSortingDemo](examples/CompoundSortingDemo)
    * A more complex example of sorting by a compound key, first by
      `score`, then breaking any ties by `name`.
* Benchmarks
    * [examples/MemoryBenchmark](examples/MemoryBenchmark)
        * Determine flash and static RAM consumption of various algorithms.
    * [examples/AutoBenchmark](examples/AutoBenchmark)
        * Determine CPU runtime of various algorithms.
    * [examples/WorstCaseBenchmark](examples/WorstCaseBenchmark)
        * Determine CPU runtime of worst case input data (e.g. sorted, reverse
          sorted).

<a name="Usage"></a>
## Usage

<a name="HeaderAndNamespace"></a>
### Include Header and Namespace

Only a single header file `AceSorting.h` is required to use this library.
To prevent name clashes with other libraries that the calling code may use, all
classes are defined in the `ace_sorting` namespace. To use the code without
prepending the `ace_sorting::` prefix, use the `using` directive to select the
specific algorithm. For example, to use the `shellSortKnut()` function, use
something like this:

```C++
#include <Arduino.h>
#include <AceSorting.h>
using ace_sorting::shellSortKnuth;
```

<a name="BubbleSort"></a>
### Bubble Sort

See https://en.wikipedia.org/wiki/Bubble_sort.

```C++
namespace ace_sorting {

template <typename T>
void bubbleSort(T data[], uint16_t n);

}
```

* Flash consumption: 44 bytes on AVR
* Additional ram consumption: none
* Runtime complexity: `O(N^2)`
* Stable sort: Yes
* Performance Notes:
    * If `data[]` is already sorted, this is `O(N)`.
    * Worst performance `O(N^2)` when data is reverse sorted.
* **Not recommended**

<a name="InsertionSort"></a>
### Insertion Sort

See https://en.wikipedia.org/wiki/Insertion_sort.

```C++
namespace ace_sorting {

template <typename T>
void insertionSort(T data[], uint16_t n);

}
```

* Flash consumption, 60 bytes on AVR
* Additional ram consumption: none
* Runtime complexity: `O(N^2)` but 5-6X faster than `bubbleSort()`
* Stable sort: Yes
* Performance Notes:
    * If `data[]` is already sorted, this algorithm is `O(N)`.
    * Worst performance `O(N^2)` when data is reverse sorted.
* **Recommendation**: Use for N smaller than about 100 and only if you need
  a stable sort

<a name="SelectionSort"></a>
### Selection Sort

See https://en.wikipedia.org/wiki/Selection_sort.

```C++
namespace ace_sorting {

template <typename T>
void selectionSort(T data[], uint16_t n);

}
```

* Flash consumption, 100 bytes on AVR
* Additional ram consumption: none
* Runtime complexity: `O(N^2)` but 2X slower than `insertionSort()`
* Stable sort: No
* Performance Notes:
    * Little change in performance when data is already sorted or reverse
      sorted.
* **Not recommended**:
    * Larger and slower than `insertionSort()` but is not a stable sort.
    * The only thing it has going for it is that it has the least number of
      writes versus reads. Might be worth looking into if writing the data is
      much more expensive than reading the data.

<a name="ShellSort"></a>
### Shell Sort

See https://en.wikipedia.org/wiki/Shellsort. Three versions are provided in this
library:

```C++
namespace ace_sorting {

template <typename T>
void shellSortClassic(T data[], uint16_t n);

template <typename T>
void shellSortKnuth(T data[], uint16_t n);

template <typename T>
void shellSortTokuda(T data[], uint16_t n);

}
```

* Flash consumption: 100-180 bytes of flash on AVR
* Additional ram consumption: none
* Runtime complexity: `O(N^k)` where `k=1.3 to 1.5`
* Stable sort: No
* Performance Notes:
    * As fast as Quick Sort for `N <= ~300`.
    * Little change in performance when data is already sorted or reverse
      sorted.
* **Recomendation**: Use `shellSortKnuth()`, which seems consistently faster
  than `shellSortClassic()`, just as fast as `shellSortTokuda()` but is simpler
  and takes less flash memory than `shellSortTokuda()`.

<a name="CombSort"></a>
### Comb Sort

See https://en.wikipedia.org/wiki/Comb_sort
and https://rosettacode.org/wiki/Sorting_algorithms/Comb_sort.
Four versions are provided in this library:

```C++
namespace ace_sorting {

template <typename T>
void combSort13(T data[], uint16_t n);

template <typename T>
void combSort13m(T data[], uint16_t n);

template <typename T>
void combSort133(T data[], uint16_t n);

template <typename T>
void combSort133m(T data[], uint16_t n);

}
```

* Flash consumption: 106-172 bytes on AVR
* Additional ram consumption: none
* Runtime complexity: `O(N^k)` where k seems similar to Shell Sort
* Stable sort: No
* Performance Notes:
    * Slightly slower than Shell Sort over similar data sets.
    * Little change in performance when data is already sorted or reverse
      sorted.
* Upper limits:
    * `combSort13()` and `combSort13m()`:  `n = 6553` on 8-bit AVR processors
      due to integer overflow.
    * `combSort133()` and `combSort133m()`: `n = 21845` on 8-bit AVR processors
      due to integer overflow.
* **Recommendation**:
    * Prefer Shell Sort over Comb Sort in most situations.
    * Use `combSort133()` on 8-bit processors.
    * Use `combSort13m()` on 32-bit processors.

The `combSort13()` and `combSort13m()` functions use a gap factor of `10/13 =
1.3`. The `combSort13m()` function modifies the gap sequence when the gap is 9
or 10. That apparently make the algorithm faster for reasons that I have not
been able to find on the internet.

The `combSort133()` function uses a gap factor of `4/3 = 1.33` instead of
`10/13`. The `4/3` ratio means that the gap size is multiplied by `3/4` on each
iteration, which allows the compiler to replace the integer division by `4` with
a right bit shift operation, so that code is smaller and faster on 8-bit
processors without hardware integer division.

The `combSort133m()` function essentially the same as `combSort133()` but
modifies the gap sequence when the gap is 9 or 10.

In terms of performance, [examples/AutoBenchmark](examples/AutoBenchmark) shows
that Comb Sort is consistently slower than Shell Sort so it is difficult to
recommend it over Shell Sort.

<a name="QuickSort"></a>
### Quick Sort

See https://en.wikipedia.org/wiki/Quicksort. Three versions are provided in this
library:

```C++
namespace ace_sorting {

template <typename T>
void quickSortMiddle(T data[], uint16_t n);

template <typename T>
void quickSortMedian(T data[], uint16_t n);

template <typename T>
void quickSortMedianSwapped(T data[], uint16_t n);

}
```

* `quickSortMiddle()`
    * The pivot is the middle element in each partition, regardless of its
      value.
* `quickSortMedian()`
    * The pivot is the median element among the 3 elements on the left, middle,
      and right slots of each partition.
* `quickSortMedianSwapped()`
    * The pivot is the median element among the 3 elements on the left, middle,
      and right slots of each partition.
    * The 3 elements are swapped so that they are sorted.
* Flash consumption: 178-278 bytes on AVR
* Additional ram consumption: `O(log(N))` bytes on stack due to recursion
* Runtime complexity: `O(N log(N))`
* Stable sort: No
* Performance Notes:
    * Fastest algorithm for large (`N >= ~1000`) data sets.
    * Little change in performance when data is already sorted or reverse
      sorted.
* **Recommendation**
    * Avoid on 8-bit processors with limited ram due to extra stack usage by
      recursion.
    * Use `quickSortMiddle()` if you have to, which is the smallest among the 3
      versions, but only slightly slower.

<a name="CLibraryQsort"></a>
### C Library Qsort

The C library `<stdlib.h>` provides a `qsort()` function that implements Quick
Sort. Since this is a standard C library function, it should be available on all
Arduino platforms. The signature looks like this:

```C++
#include <stdlib.h>

void qsort(void *base, size_t nmemb, size_t size,
    int (*compar)(const void *, const void *));
```

It has the following characteristics:

* Flash consumption: 1084 bytes on AVR
* Additional ram consumption: `O(log(N))` bytes on stack due to recursion
* Runtime complexity: `O(N log(N))`, but 2-3X slower than C++ versions in this
  library
* Stable sort: No
* **Not recommended** due to excessive flash consumption and slowness.

According to benchmarks, `qsort()` is 2-3X slower than the C++ `quickSortXxx()`
functions provided by this library, and consumes 4-5X more flash memory. The
`qsort()` function is probably more sophisticated in the handling of edge cases,
but it suffers from being a general function that uses a call-back function. The
overhead of that function call makes it 2-3X slower than the C++ template
functions in this library where the comparison function call is usually inlined
by the compiler. For these reasons, it is difficult to recommend the C-library
`qsort()` function.

<a name="AdvancedUsage"></a>
## Advanced Usage

Each sorting algorithm comes in another variant that takes 3 arguments instead
of 2 arguments. For completeness, here are the signatures of the 3-argument
variants:

```C++
namespace ace_sorting {

template <typename T, typename F>
void bubbleSort(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void insertionSort(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void selectionSort(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void shellSortClassic(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void shellSortKnuth(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void shellSortTokuda(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void combSort13(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void combSort13m(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void combSort133(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void combSort133m(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void quickSortMiddle(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void quickSortMedian(T data[], uint16_t n, F&& lessThan);

template <typename T, typename F>
void quickSortMedianSwapped(T data[], uint16_t n, F&& lessThan);

}
```

The 3rd argument `lessThan` is a user-defined function pointer or a lambda
expression that implements the "less than" comparison operator between 2
elements in the `data[]` array. (The `F&&` is a C++11 syntax that means "rvalue
reference to a templatized type of `F`".) In the context of this library, the
`lessThan` parameter can be thought of as a function-like thing that accepts 2
elements of type `T` and returns a `bool`. Two types of this `lessThan`
parameter is explained below.

<a name="FunctionPointer"></a>
### Function Pointer

If `lessThan` is a function pointer, then its signature should look something
like one of these:

```C++
bool lessThan(const T& a, const T& b);

bool lessThan(T a, T b);
```

The [examples/CompoundSortingDemo](examples/CompoundSortingDemo) example shows
what this looks like to sort an array of pointers to `Record` entries by
`score`, then by `name`, to break any ties:

```C++
struct Record {
  const char* name;
  int score;
};

bool sortByScoreThenName(const Record* a, const Record* b) {
  if (a->score < b->score) {
    return true;
  } else if (a->score > b->score) {
    return false;
  } else {
    return strcmp(a->name, b->name) < 0;
  }
}

const uint16_t ARRAY_SIZE = 10;
const Record RECORDS[ARRAY_SIZE] = {
  ...
};

const Record* recordPtrs[ARRAY_SIZE];

void doSorting() {
  shellSortKnuth(recordPtrs, ARRAY_SIZE, sortByScoreThenName);
  ...
}
```

<a name="LambdaExpression"></a>
### Lambda Expression

Through the magic of C++ templates, the `lessThan` parameter can also be a
lambda expression. When the comparison operator is simple and short, this
feature can save us the hassle of creating a separate function that is used only
once.

The [examples/HelloSorting](examples/HelloSorting) example shows how to sort an
array of integers in reverse order using an inlined lambda expression that
reverses the comparison operator:

```C++
const uint16_t ARRAY_SIZE = 20;
int array[ARRAY_SIZE];

void doSorting() {
  shellSortKnuth(array, ARRAY_SIZE, [](int a, int b) { return a > b; });
  ...
}
```

<a name="CompilerOptimizations"></a>
### Compiler Optimizations

All 2-argument variants of the sorting functions (except for the Quick Sort
functions `quickSortXxx()`, see below) are implemented by simply calling the
3-argument sorting functions using a default lambda expression using the `<`
operator like this:

```C++
template <typename T>
void shellSortKnuth(T data[], uint16_t n) {
  auto&& lessThan = [](const T& a, const T& b) -> bool { return a < b; };
  shellSortKnuth(data, n, lessThan);
}
```

The benchmarks in [examples/MemoryBenchmark](examples/MemoryBenchmark) show that
the compiler is able to completely optimize away the overhead of the inlined
lambda expression and produce code that is equivalent to inserting the `<`
binary operator directly into the code that implements the 3-argument variant.
No additional flash memory is consumed.

The only exception to this compiler optimization occurs with the Quick Sort
algorithms. The compiler seems unable to optimize away the extra layer of
indirection, probably due to the recursive function calls in the Quick Sort
algorithms. Therefore, the 2-argument variants of `quickSortXxx()` algorithms
are duplicated from the 3-argument variants with the simple `<` operator
hardcoded, so that the simple case of ascending sorting consumes as little flash
memory as possible. (In the source code, the `ACE_SORTING_DIRECT_QUICK_SORT`
macro is set to `1` by default to achieve this, in contrast to all other sorting
functions where the equivalent macro is set to `0`.)

<a name="ResourceConsumption"></a>
## Resource Consumption

<a name="FlashAndStaticMemory"></a>
### Flash And Static Memory

The full details of flash and static memory consumptions are given in
[examples/MemoryBenchmark](examples/MemoryBenchmark). Here are 2 samples.

**SparkFun Pro Micro (ATmega32U4)**

```
+---------------------------------------------------------------------+
| Functionality                          |  flash/  ram |       delta |
|----------------------------------------+--------------+-------------|
| Baseline                               |   4060/  354 |     0/    0 |
|----------------------------------------+--------------+-------------|
| bubbleSort()                           |   4104/  354 |    44/    0 |
| insertionSort()                        |   4120/  354 |    60/    0 |
| selectionSort()                        |   4148/  354 |    88/    0 |
|----------------------------------------+--------------+-------------|
| shellSortClassic()                     |   4156/  354 |    96/    0 |
| shellSortKnuth()                       |   4202/  354 |   142/    0 |
| shellSortTokuda()                      |   4242/  380 |   182/   26 |
|----------------------------------------+--------------+-------------|
| combSort13()                           |   4214/  354 |   154/    0 |
| combSort13m()                          |   4232/  354 |   172/    0 |
| combSort133()                          |   4166/  354 |   106/    0 |
| combSort133m()                         |   4188/  354 |   128/    0 |
|----------------------------------------+--------------+-------------|
| quickSortMiddle()                      |   4238/  354 |   178/    0 |
| quickSortMedian()                      |   4290/  354 |   230/    0 |
| quickSortMedianSwapped()               |   4336/  354 |   276/    0 |
|----------------------------------------+--------------+-------------|
| qsort()                                |   5144/  354 |  1084/    0 |
+---------------------------------------------------------------------+
```

**ESP8266**

```
+---------------------------------------------------------------------+
| Functionality                          |  flash/  ram |       delta |
|----------------------------------------+--------------+-------------|
| Baseline                               | 260497/28108 |     0/    0 |
|----------------------------------------+--------------+-------------|
| bubbleSort()                           | 260545/28108 |    48/    0 |
| insertionSort()                        | 260545/28108 |    48/    0 |
| selectionSort()                        | 260561/28108 |    64/    0 |
|----------------------------------------+--------------+-------------|
| shellSortClassic()                     | 260577/28108 |    80/    0 |
| shellSortKnuth()                       | 260577/28108 |    80/    0 |
| shellSortTokuda()                      | 260637/28128 |   140/   20 |
|----------------------------------------+--------------+-------------|
| combSort13()                           | 260609/28108 |   112/    0 |
| combSort13m()                          | 260625/28108 |   128/    0 |
| combSort133()                          | 260593/28108 |    96/    0 |
| combSort133m()                         | 260609/28108 |   112/    0 |
|----------------------------------------+--------------+-------------|
| quickSortMiddle()                      | 260641/28108 |   144/    0 |
| quickSortMedian()                      | 260673/28108 |   176/    0 |
| quickSortMedianSwapped()               | 260689/28108 |   192/    0 |
|----------------------------------------+--------------+-------------|
| qsort()                                | 261617/28108 |  1120/    0 |
+---------------------------------------------------------------------+
```

<a name="CpuCycles"></a>
### CPU Cycles

The CPU benchmark numbers can be seen in
[examples/AutoBenchmark](examples/AutoBenchmark). All times in milliseconds.
Here are 2 samples.

**SparkFun Pro Micro (ATmega32U4)**

```
+---------------------+------------------------+---------+---------+---------+
|            \      N |    10 |    30 |    100 |     300 |    1000 |    3000 |
| Function    \       |       |       |        |         |         |         |
|---------------------+-------+-------+--------+---------+---------+---------|
| bubbleSort()        | 0.083 | 1.211 | 13.093 | 118.403 |         |         |
| insertionSort()     | 0.044 | 0.259 |  2.515 |  21.941 | 240.911 |         |
| selectionSort()     | 0.084 | 0.555 |  5.388 |  46.517 | 509.037 |         |

CHANGELOG.md
# Changelog

* Unreleased
* v1.0.0 (2021-12-04)
    * Add special case for SparkFun Pro Micro in `examples/AutoBenchmark` whose
      ATmega32U4 has 2.5kB of ram, enough to handle `N=1000`.
    * Fix incorrect C++ header guard in `src/AceSorting.h`.
    * Upgrade tool chain
        * Upgrade Arduino IDE to 1.8.16.
        * Upgrade Arduino CLI to 0.19.2.
        * Upgrade SparkFun SAMD to 1.8.4.
        * Upgrade ESP8266 to 3.0.2.
        * Upgrade TeensyDuino to 1.55.
* v0.3 (2021-08-23)
    * Add 3-argument versions of each sorting function which accepts a lambda
      expression or function pointer that evalutes the general "less-than"
      comparison between 2 elements in the array.
        * Can be used to implement any sorting criteria, e.g. reverse sort, or
          sorting using compound keys.
        * Reimplement 2-argument versions of each sorting function to simply
          call the 3-argument versions using a default "less-than" function
          to give simple ascending order.
        * The `quickSortXxx()` algorithms do not optimize well, probably due
          to the recursive function call, so the 2-argument version is
          duplicated from the 3-argument version.
        * Add `examples/CompoundSortingDemo` to illustrate the 3-argument
          sorting functions to sort by (name, score) pair.
        * No changes observed to flash memory consumption or CPU execution
          times.
* v0.2 (2021-08-06)
    * Add Selection Sort, mostly for completeness. It's another `O(N^2)` sort
      but is slower than Insertion Sort, and is not a stable sort.
    * Add [examples/HelloSorting](examples/HelloSorting).
    * Add **tl;dr** section in README.md to summarize my recommendations.
* v0.1 (2021-08-04)
    * Add `combSort133()` which uses a gap factor of 4/3, which eliminates
      integer division. Smaller and faster on 8-bit processors which don't have
      hardware integer division.
    * Add `combSort13m()` and `combSort133m()` which modifies the gap to 11 when
      equal to 9 and 10.
    * Add `WorstCaseBenchmark` to determine behavior at extremes (already
      sorted, and reversed sorted).
    * Add comparison to C library's builtin `qsort()`. Seems like it's 2-3X
      slower and consumes 4-5X more flash memory than `quickSortXxx()`.
    * Add unit tests.
* Initial checking (2021-08-02)


examples/AutoBenchmark/Benchmark.cpp
/*
 * MIT License
 * Copyright (c) 2021 Brian T. Park
 */

#include <stdint.h> // uint8_t, uint16_t
#include <stdlib.h> // qsort()
#include <Arduino.h> // F(), __FlashStringHelper
#include <AceCommon.h>
#include <AceSorting.h>
#include "Benchmark.h"

#ifndef SERIAL_PORT_MONITOR
  #define SERIAL_PORT_MONITOR Serial
#endif

using ace_common::GenericStats;
using ace_common::isSorted;
using ace_sorting::bubbleSort;
using ace_sorting::insertionSort;
using ace_sorting::selectionSort;
using ace_sorting::shellSortClassic;
using ace_sorting::shellSortKnuth;
using ace_sorting::shellSortTokuda;
using ace_sorting::combSort13;
using ace_sorting::combSort13m;
using ace_sorting::combSort133;
using ace_sorting::combSort133m;
using ace_sorting::quickSortMiddle;
using ace_sorting::quickSortMedian;
using ace_sorting::quickSortMedianSwapped;

//-----------------------------------------------------------------------------
// Benchmark parameters.
//-----------------------------------------------------------------------------

#if defined(EPOXY_DUINO)
const uint16_t DATA_SIZES[] = {10, 30, 100, 300, 1000, 3000, 10000, 30000};
//const uint16_t DATA_SIZES[] = {65000};
const uint16_t SLOW_SAMPLE_SIZE = 3;
const uint16_t FAST_SAMPLE_SIZE = 25;
#elif defined(ARDUINO_AVR_PROMICRO)
// ATmega32U4 has 2.5kB of ram, so it can handle N=1000... except qsort()
// which seems to run out of stack space.
const uint16_t DATA_SIZES[] = {10, 30, 100, 300, 1000};
const uint16_t SLOW_SAMPLE_SIZE = 3;
const uint16_t FAST_SAMPLE_SIZE = 25;
#elif defined(ARDUINO_ARCH_AVR)
const uint16_t DATA_SIZES[] = {10, 30, 100, 300};
const uint16_t SLOW_SAMPLE_SIZE = 3;
const uint16_t FAST_SAMPLE_SIZE = 25;
#else
// All 32-bit boards that I own can handle at least N=3000.
const uint16_t DATA_SIZES[] = {10, 30, 100, 300, 1000, 3000};
const uint16_t SLOW_SAMPLE_SIZE = 3;
const uint16_t FAST_SAMPLE_SIZE = 20;
#endif
const uint16_t NUM_DATA_SIZES = sizeof(DATA_SIZES) / sizeof(DATA_SIZES[0]);

//-----------------------------------------------------------------------------
// Stats helpers
//-----------------------------------------------------------------------------

/**
 * Print the result in micros for the given 'name' function or algorithm. The
 * format is:
 *
 * @verbatim
 * name min avg max sampleSize
 * @endverbatim
 *
 * @param name name of the benchmark
 * @param instance of GenericStats
 * @param samplesize number of independent timingStats samples used, to get
 *        the min(), avg() and max()
 */
static void printStats(
    const __FlashStringHelper* name,
    const GenericStats<float>& stats,
    uint16_t sampleSize,
    uint16_t dataSize) {
  SERIAL_PORT_MONITOR.print(name);
  SERIAL_PORT_MONITOR.print(' ');
  SERIAL_PORT_MONITOR.print(dataSize);
  SERIAL_PORT_MONITOR.print(' ');
  SERIAL_PORT_MONITOR.print(stats.getMin(), 3);
  SERIAL_PORT_MONITOR.print(' ');
  SERIAL_PORT_MONITOR.print(stats.getAvg(), 3);
  SERIAL_PORT_MONITOR.print(' ');
  SERIAL_PORT_MONITOR.print(stats.getMax(), 3);
  SERIAL_PORT_MONITOR.print(' ');
  SERIAL_PORT_MONITOR.println(sampleSize);
}

static GenericStats<float> timingStats;

//-----------------------------------------------------------------------------
// Benchmark helpers
//-----------------------------------------------------------------------------

// Typedef of the lambda expression that can be converted into a simple function
// pointer.
typedef void (*SortFunction)(uint16_t data[], uint16_t n);

// Volatile to prevent the compiler from deciding that the entire program does
// nothing, causing it to opimize the whole program to nothing.
volatile uint32_t disableCompilerOptimization;

// Pointer to the data array created by malloc() and free(). This is a global
// variable to avoid using capture in the lambda expression, so that it can
// degenerate into a function pointer.
uint16_t* array;

static void fillArray(uint16_t data[], uint16_t n) {
  for (uint16_t i = 0; i < n; ++i) {
    data[i] = random(65536);
  }
}

static int compare(const void* a, const void* b) {
  uint16_t va = *((uint16_t*) a);
  uint16_t vb = *((uint16_t*) b);
  return (va < vb) ? -1 : ((va == vb) ? 0 : 1);
}

static void doQsort(uint16_t data[], uint16_t n) {
  qsort(data, n, sizeof(uint16_t), compare);
}

static void runSort(
    const __FlashStringHelper* name,
    uint16_t dataSize,
    uint16_t sampleSize,
    SortFunction sortFunction) {

  timingStats.reset();
  array = new uint16_t[dataSize];

  for (uint8_t k = 0; k < sampleSize; k++) {
    fillArray(array, dataSize);
    bool issorted = isSorted(array, dataSize);
    if (issorted) {
      SERIAL_PORT_MONITOR.println(F("Original array is unexpectedly sorted"));
    }

    yield();
    uint32_t startMicros = micros();
    sortFunction(array, dataSize);
    uint32_t elapsedMicros = micros() - startMicros;
    yield();
    disableCompilerOptimization = array[0];

    issorted = isSorted(array, dataSize);
    if (! issorted) {
      SERIAL_PORT_MONITOR.println(F("Sorted array is NOT sorted!"));
    }
    timingStats.update((float) elapsedMicros / 1000.0);
  }

  delete[] array;
  printStats(name, timingStats, sampleSize, dataSize);
}

static void runSortForSizes(
    const __FlashStringHelper* name,
    uint16_t sampleSize,
    SortFunction sortFunction) {
  for (uint16_t i = 0; i < NUM_DATA_SIZES; i++) {
    uint16_t dataSize = DATA_SIZES[i];

    // Don't run O(N^2) sorting algorithms on large arrays because they take too
    // long to finish.
    if (sortFunction == &bubbleSort<uint16_t>
        || sortFunction == &insertionSort<uint16_t>
        || sortFunction == &selectionSort<uint16_t>) {
      if (dataSize > 1000) break;
    }
    // Don't run qsort for N>=1000 on Pro Micro because it runs out of stack
    // space.
    #if defined(ARDUINO_AVR_PROMICRO)
      if (sortFunction == doQsort && dataSize >= 1000) break;
    #endif

    // Don't run bubbleSort() with N>=1000 any AVR because it takes too long.
    #if defined(ARDUINO_ARCH_AVR)
      if (sortFunction == &bubbleSort<uint16_t> && dataSize >= 1000) break;
    #endif

    runSort(name, dataSize, sampleSize, sortFunction);
  }
}

//-----------------------------------------------------------------------------
// Benchmarks
//-----------------------------------------------------------------------------

void runBenchmarks() {
#if ! defined(EPOXY_DUINO)
  runSortForSizes(
      F("bubbleSort()"), SLOW_SAMPLE_SIZE, bubbleSort<uint16_t>);
#endif
  runSortForSizes(
      F("insertionSort()"), SLOW_SAMPLE_SIZE, insertionSort<uint16_t>);
  runSortForSizes(
      F("selectionSort()"), SLOW_SAMPLE_SIZE, selectionSort<uint16_t>);

  runSortForSizes(
      F("shellSortClassic()"), FAST_SAMPLE_SIZE, shellSortClassic<uint16_t>);
  runSortForSizes(
      F("shellSortKnuth()"), FAST_SAMPLE_SIZE, shellSortKnuth<uint16_t>);
  runSortForSizes(
      F("shellSortTokuda()"), FAST_SAMPLE_SIZE, shellSortTokuda<uint16_t>);

  runSortForSizes(
      F("combSort13()"), FAST_SAMPLE_SIZE, combSort13<uint16_t>);
  runSortForSizes(
      F("combSort13m()"), FAST_SAMPLE_SIZE, combSort13m<uint16_t>);
  runSortForSizes(
      F("combSort133()"), FAST_SAMPLE_SIZE, combSort133<uint16_t>);
  runSortForSizes(
      F("combSort133m()"), FAST_SAMPLE_SIZE, combSort133m<uint16_t>);

  runSortForSizes(
      F("quickSortMiddle()"), FAST_SAMPLE_SIZE, quickSortMiddle<uint16_t>);
  runSortForSizes(
      F("quickSortMedian()"), FAST_SAMPLE_SIZE, quickSortMedian<uint16_t>);
  runSortForSizes(
      F("quickSortMedianSwapped()"),
      FAST_SAMPLE_SIZE,
      quickSortMedianSwapped<uint16_t>);
  runSortForSizes(
      F("qsort()"), FAST_SAMPLE_SIZE, doQsort);
}
